### ISSUE TITLE
windowscodecs not linked.

### ISSUE THREAD

**Issue by API-Beast** [dacap/clip#35]
clip_win_wic.h uses functions from the "windowscodecs" library, but this isn't linked in the CMakeLists.txt resulting in "unresolved symbol" errors.

Tested on Windows with GCC 10.2.0.

**Comment by dacap** [comment]
Hi @API-Beast, what are those specific unresolved symbols (what is the specific output of gcc)? Because the clip_win_wic.h header uses COM objects and those are dynamically linked (COM vtables are just defined with structs in .h files, so there is no need of an extra .lib). I guess that maybe some GUID is defined in a .a library in the gcc/mingw case (e.g. for `GUID_WICPixelFormat32bppBGRA`, etc.) but I'm not sure.

**Comment by API-Beast** [comment]
```bash
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: clip/libclip.a(clip_win.cpp.obj):clip_win.cpp:(.rdata$.refptr.CLSID_WICPngDecoder1[.refptr.CLSID_WICPngDecoder1]+0x0): undefined reference to `CLSID_WICPngDecoder1'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: clip/libclip.a(clip_win.cpp.obj):clip_win.cpp:(.rdata$.refptr.CLSID_WICPngDecoder2[.refptr.CLSID_WICPngDecoder2]+0x0): undefined reference to `CLSID_WICPngDecoder2'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: clip/libclip.a(clip_win.cpp.obj):clip_win.cpp:(.rdata$.refptr.GUID_WICPixelFormat32bppBGRA[.refptr.GUID_WICPixelFormat32bppBGRA]+0x0): undefined reference to `GUID_WICPixelFormat32bppBGRA'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: clip/libclip.a(clip_win.cpp.obj):clip_win.cpp:(.rdata$.refptr.CLSID_WICPngEncoder[.refptr.CLSID_WICPngEncoder]+0x0): undefined reference to `CLSID_WICPngEncoder'
```

WICPngDecoder1, WICPngDecoder2, WICPixelFormat32bppBGRA and WICPngEncoder

**Comment by dacap** [comment]
As I was suspecting, it's related to the undefined GUIDs/CLSIDs. Not sure where they are defined in the MinGW implementation.

**Comment by API-Beast** [comment]
Hmm, at the very least I can confirm that linking windowscodecs resolves those symbols.

**Comment by agudpp** [comment]
@dacap I'm having some similar issues, letting the problem here since it is related, I'm using `mingw-w64\x86_64-8.1.0-posix-seh-rt_v6-rev0` on windows.
Neither `CLSID_WICPngDecoder2` nor `CLSID_WICPngDecoder1` are in `#include <wincodec.h>` so it doesn't compile:
```cpp
HRESULT WINAPI WICConvertBitmapSource(REFWICPixelFormatGUID dstFormat, IWICBitmapSource *pISrc, IWICBitmapSource **ppIDst);
DEFINE_GUID(CLSID_WICBmpDecoder, 0x6b462062,0x7cbf,0x400d,0x9f,0xdb,0x81,0x3d,0xd1,0x0f,0x27,0x78);
DEFINE_GUID(CLSID_WICPngDecoder, 0x389ea17b,0x5078,0x4cde,0xb6,0xef,0x25,0xc1,0x51,0x75,0xc7,0x51);
DEFINE_GUID(CLSID_WICIcoDecoder, 0xc61bfcdf,0x2e0f,0x4aad,0xa8,0xd7,0xe0,0x6b,0xaf,0xeb,0xcd,0xfe);
// ....
DEFINE_GUID(CLSID_WICPngEncoder, 0x27949969,0x876a,0x41d7,0x94,0x47,0x56,0x8f,0x6a,0x35,0xa4,0xdc);
```
I have to change the code to use `CLSID_WICPngDecoder` in order to compile. The lib compiles fine, but the tests fails with the similar errors shown above:
```bash
[ 40%] Built target clip
[ 50%] Linking CXX executable clip_user_format_tests.exe
../libclip.a(clip_win.cpp.obj):clip_win.cpp:(.rdata$.refptr.CLSID_WICPngDecoder[.refptr.CLSID_WICPngDecoder]+0x0): undefined reference to `CLSID_WICPngDecoder'
../libclip.a(clip_win.cpp.obj):clip_win.cpp:(.rdata$.refptr.GUID_WICPixelFormat32bppBGRA[.refptr.GUID_WICPixelFormat32bppBGRA]+0x0): undefined reference to `GUID_WICPixelFormat32bppBGRA'
../libclip.a(clip_win.cpp.obj):clip_win.cpp:(.rdata$.refptr.CLSID_WICPngEncoder[.refptr.CLSID_WICPngEncoder]+0x0): undefined reference to `CLSID_WICPngEncoder'
collect2.exe: error: ld returned 1 exit status
mingw32-make.exe[2]: *** [tests\CMakeFiles\clip_user_format_tests.dir\build.make:88: tests/clip_user_format_tests.exe] Error 1
mingw32-make.exe[1]: *** [CMakeFiles\Makefile2:126: tests/CMakeFiles/clip_user_format_tests.dir/all] Error 2
mingw32-make.exe: *** [Makefile:94: all] Error 2
```

I'm using your latest change:
```
  # MinGW requires the windowscodecs just because CLSIDs are defined
  # in the windowscodecs.a file instead of the wincodec.h file (?!)
  if(MINGW)
    find_library(CLIP_WINDOWSCODECS_LIBRARY windowscodecs)
    if(CLIP_WINDOWSCODECS_LIBRARY)
      target_link_libraries(clip ${CLIP_WINDOWSCODECS_LIBRARY})
    endif()
  endif()
```
But seems to not work still... I will be researching, but if any clue will be great :)

**Comment by dacap** [comment]
Hi @agudpp! It looks that you have an older version mingw64-headers (link of [latest wincodec.h](https://github.com/mirror/mingw-w64/blob/12b5d1344c23c5fdba7fb2c771d0476c2220db16/mingw-w64-headers/include/wincodec.h#L6185)).

Anyway I cannot help too much because I wasn't able to test this change (I don't have the latest MinGW 64 version, available, nor sure where to find it, also I'm not sure what is official or a distro that can be trusted).

**Comment by API-Beast** [comment]
Currently the best way to install MinGW is via [MSYS2](https://www.msys2.org/).

**Comment by agudpp** [comment]
hi @dacap thanks for the fast reply. Interestingly I assumed was the latest version of mingw since I downloaded from the official website.
I will check it with a new version I'm downloading from Qt and let you know!

cool project btw :)

**Comment by KyrietS** [comment]
I have the same problem with the latest MinGW-w64. I changed this:
```cmake
  if(MINGW)
    find_library(CLIP_WINDOWSCODECS_LIBRARY windowscodecs)
    if(CLIP_WINDOWSCODECS_LIBRARY)
      target_link_libraries(clip ${CLIP_WINDOWSCODECS_LIBRARY})
    endif()
  endif()
```

to this:
```cmake
  if(MINGW)
    target_link_libraries(clip windowscodecs)
  endif()
```

And now it links and it compiles. Any idea why `find_library` cannot find the library? Maybe you should skip the CMake's searching part and just let the compiler find it assuming that on Windows platform this library is always be available?

**Comment by KyrietS** [comment]
Hi, sure. I even added some messages:
```cmake
  if(MINGW)
    message("lizard looking for windowscodecs")
    find_library(CLIP_WINDOWSCODECS_LIBRARY windowscodecs)
    message(${CLIP_WINDOWSCODECS_LIBRARY})
    if(CLIP_WINDOWSCODECS_LIBRARY)
      target_link_libraries(clip ${CLIP_WINDOWSCODECS_LIBRARY})
    endif()
  endif()
```

And when I ran `cmake --debug-output ..` (in CLion this command is a little bit longer) I got nothing printed:
```
Running with debug output on.
lizard looking for windowscodecs
   Called from: [1]	C:/Users/Me/clip/CMakeLists.txt
CLIP_WINDOWSCODECS_LIBRARY-NOTFOUND
```

Generally I am a little bit lost with this "MinGW" stuff. It seems way too complicated for me 😅. I just want to open your library in CLion with MinGW toolset and compile the clip's examples. But it doesn't work out-of-the-box 😢

Your CI job proves that if I had MingGW configured in my system "correctly" then it should work. But I can't do:
```yaml
 - uses: msys2/setup-msys2@v2
```

On my system, like you do in your CI, unfortunately 😄
I suspect that I may use CMake from "Windows" and GCC compiler from "MinGW". That's why CMake can't find the library (because it searches in Windows' locations), but the linker has no problem with that. I have no idea how to fix it though ¯\\_(ツ)_/¯

Have you ever tried to compile the examples in CLion with MinGW toolset?
